File: app/src/main/java/com/wineguesser/deductive/view/adapter/ConclusionItemAdapter.java
package com.wineguesser.deductive.view.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.wineguesser.deductive.R;
import com.wineguesser.deductive.model.ConclusionRecord;


import java.util.List;

public class ConclusionItemAdapter extends RecyclerView.Adapter<ConclusionItemAdapter.ConclusionItemAdapterViewHolder> {

    private final Context mContext;
    private List<ConclusionRecord> mConclusionRecords;
    private final HistoryItemOnClickHandler mClickHandler;

    public interface HistoryItemOnClickHandler {
        void onHistoryItemClick(ConclusionRecord conclusionRecord);
    }

    public ConclusionItemAdapter(Context context, HistoryItemOnClickHandler clickHandler) {
        mContext = context;
        mClickHandler = clickHandler;
    }

    class ConclusionItemAdapterViewHolder extends RecyclerView.ViewHolder
    implements View.OnClickListener {

        final TextView mActualLabel;
        final TextView mActualGrape;
        final TextView mUserGrape;
        final TextView mAppGrape;

        ConclusionItemAdapterViewHolder(View view) {
            super(view);

            mActualLabel = view.findViewById(R.id.textView_actual_label);
            mActualGrape = view.findViewById(R.id.textView_actual_variety);
            mUserGrape = view.findViewById(R.id.textView_user_conclusion_grape);
            mAppGrape = view.findViewById(R.id.textView_app_conclusion_grape);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            ConclusionRecord conclusionRecord = mConclusionRecords.get(adapterPosition);
            mClickHandler.onHistoryItemClick(conclusionRecord);
        }
    }

    @NonNull
    public ConclusionItemAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        int layoutId = R.layout.item_conclusion_record;
        LayoutInflater inflater = LayoutInflater.from(mContext);

        View view = inflater.inflate(layoutId, viewGroup, false);
        view.setFocusable(true);

        return new ConclusionItemAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ConclusionItemAdapterViewHolder conclusionItemAdapterViewHolder, int position) {

        conclusionItemAdapterViewHolder.mActualLabel.setText(mConclusionRecords.get(position).getActualLabel());

        conclusionItemAdapterViewHolder.mActualGrape.setText(mConclusionRecords.get(position).getActualVariety());
        conclusionItemAdapterViewHolder.mUserGrape.setText(mConclusionRecords.get(position).getUserConclusionVariety());
        String appConclusionVariety = mConclusionRecords.get(position).getAppConclusionVariety();
        if (appConclusionVariety == null || appConclusionVariety.isEmpty()) {
            conclusionItemAdapterViewHolder.mAppGrape.setText(R.string.no_data);
        } else {
            conclusionItemAdapterViewHolder.mAppGrape.setText(appConclusionVariety);
        }
    }

    @Override
    public int getItemCount() {
        if (mConclusionRecords == null) return 0;
        return mConclusionRecords.size();
    }

    public void setConclusionData(List<ConclusionRecord> conclusionData) {
        mConclusionRecords = conclusionData;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/wineguesser/deductive/view/adapter/ConclusionItemAdapter.kt
package com.wineguesser.deductive.view.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.wineguesser.deductive.R
import com.wineguesser.deductive.model.ConclusionRecord

class ConclusionItemAdapter(
    private val mContext: Context,
    private val mClickHandler: HistoryItemOnClickHandler
) : RecyclerView.Adapter<ConclusionItemAdapter.ConclusionItemAdapterViewHolder>() {

    private var mConclusionRecords: List<ConclusionRecord>? = null

    interface HistoryItemOnClickHandler {
        fun onHistoryItemClick(conclusionRecord: ConclusionRecord)
    }

    inner class ConclusionItemAdapterViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        val mActualLabel: TextView = view.findViewById(R.id.textView_actual_label)
        val mActualGrape: TextView = view.findViewById(R.id.textView_actual_variety)
        val mUserGrape: TextView = view.findViewById(R.id.textView_user_conclusion_grape)
        val mAppGrape: TextView = view.findViewById(R.id.textView_app_conclusion_grape)

        init {
            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val adapterPosition = adapterPosition
            if (adapterPosition != RecyclerView.NO_POSITION) {
                mConclusionRecords?.get(adapterPosition)?.let {
                    mClickHandler.onHistoryItemClick(it)
                }
            }
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ConclusionItemAdapterViewHolder {
        val layoutId = R.layout.item_conclusion_record
        val inflater = LayoutInflater.from(mContext)
        val view = inflater.inflate(layoutId, viewGroup, false)
        view.isFocusable = true
        return ConclusionItemAdapterViewHolder(view)
    }

    override fun onBindViewHolder(holder: ConclusionItemAdapterViewHolder, position: Int) {
        val record = mConclusionRecords?.get(position) ?: return

        holder.mActualLabel.text = record.actualLabel
        holder.mActualGrape.text = record.actualVariety
        holder.mUserGrape.text = record.userConclusionVariety

        val appConclusionVariety = record.appConclusionVariety
        if (appConclusionVariety.isNullOrEmpty()) {
            holder.mAppGrape.setText(R.string.no_data)
        } else {
            holder.mAppGrape.text = appConclusionVariety
        }
    }

    override fun getItemCount(): Int {
        return mConclusionRecords?.size ?: 0
    }

    fun setConclusionData(conclusionData: List<ConclusionRecord>?) {
        mConclusionRecords = conclusionData
        notifyDataSetChanged()
    }
}
